File: pkg/reactive/handler.go
// Copyright 2025, Command Line Inc.
// SPDX-License-Identifier: Apache-2.0

package reactive

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"log"
	"strings"
	"sync"
	"time"

	"github.com/a5af/agentmux/pkg/waveobj"
	"github.com/a5af/agentmux/pkg/wstore"
	"github.com/google/uuid"
)

// InputSender is a function type that sends input to a block's PTY.
// This decouples the handler from direct blockcontroller dependency.
type InputSender func(blockId string, inputData []byte) error

// Handler manages reactive message injection for agent-to-agent communication.
type Handler struct {
	mu sync.RWMutex

	// agentToBlock maps agent IDs to block IDs
	agentToBlock map[string]string

	// blockToAgent is the reverse mapping
	blockToAgent map[string]string

	// agentInfo stores detailed registration info
	agentInfo map[string]*AgentRegistration

	// inputSender is the function to send input to blocks
	inputSender InputSender

	// auditLog stores recent injection attempts (ring buffer)
	auditLog     []AuditLogEntry
	auditLogSize int
	auditLogIdx  int

	// config
	includeSourceInMessage bool

	// Rate limiting (protects against DoS while keeping synchronous injection)
	rateLimitMu      sync.Mutex
	rateLimitTokens  int
	rateLimitMax     int
	rateLimitRefresh time.Time
}

// NewHandler creates a new reactive message handler.
func NewHandler(inputSender InputSender) *Handler {
	return &Handler{
		agentToBlock:           make(map[string]string),
		blockToAgent:           make(map[string]string),
		agentInfo:              make(map[string]*AgentRegistration),
		inputSender:            inputSender,
		auditLog:               make([]AuditLogEntry, 100), // Keep last 100 entries
		auditLogSize:           100,
		auditLogIdx:            0,
		includeSourceInMessage: false, // Can be configured
		rateLimitMax:           10,    // 10 injections per second max
		rateLimitTokens:        10,
		rateLimitRefresh:       time.Now(),
	}
}

// checkRateLimit returns true if the request is allowed, false if rate limited.
// Uses a simple token bucket: 10 tokens per second, refilled each second.
func (h *Handler) checkRateLimit() bool {
	h.rateLimitMu.Lock()
	defer h.rateLimitMu.Unlock()

	now := time.Now()
	// Refill tokens every second
	if now.Sub(h.rateLimitRefresh) >= time.Second {
		h.rateLimitTokens = h.rateLimitMax
		h.rateLimitRefresh = now
	}

	if h.rateLimitTokens > 0 {
		h.rateLimitTokens--
		return true
	}
	return false
}

// RegisterAgent associates an agent ID with a block ID.
// Called when a pane's shell sets WAVEMUX_AGENT_ID.
func (h *Handler) RegisterAgent(agentID, blockID, tabID string) error {
	// Normalize agent ID to lowercase for case-insensitive matching
	agentID = strings.ToLower(agentID)

	if !ValidateAgentID(agentID) {
		return fmt.Errorf("invalid agent ID: %s", agentID)
	}

	h.mu.Lock()
	defer h.mu.Unlock()

	// Check if this block already has a different agent
	if existingAgent, ok := h.blockToAgent[blockID]; ok && existingAgent != agentID {
		// Unregister the old agent
		delete(h.agentToBlock, existingAgent)
		delete(h.agentInfo, existingAgent)
	}

	// Check if this agent is already registered to a different block
	if existingBlock, ok := h.agentToBlock[agentID]; ok && existingBlock != blockID {
		// Unregister from the old block
		delete(h.blockToAgent, existingBlock)
	}

	now := time.Now()
	h.agentToBlock[agentID] = blockID
	h.blockToAgent[blockID] = agentID
	h.agentInfo[agentID] = &AgentRegistration{
		AgentID:      agentID,
		BlockID:      blockID,
		TabID:        tabID,
		RegisteredAt: now,
		LastSeen:     now,
	}

	log.Printf("[reactive] registered agent %s -> block %s", agentID, blockID)
	return nil
}

// UnregisterAgent removes an agent's registration.
// Called when a pane closes or agent ID is cleared.
func (h *Handler) UnregisterAgent(agentID string) {
	h.mu.Lock()
	defer h.mu.Unlock()

	if blockID, ok := h.agentToBlock[agentID]; ok {
		delete(h.blockToAgent, blockID)
	}
	delete(h.agentToBlock, agentID)
	delete(h.agentInfo, agentID)

	log.Printf("[reactive] unregistered agent %s", agentID)
}

// UnregisterBlock removes registration by block ID.
// Called when a block/pane is closed.
func (h *Handler) UnregisterBlock(blockID string) {
	h.mu.Lock()
	defer h.mu.Unlock()

	if agentID, ok := h.blockToAgent[blockID]; ok {
		delete(h.agentToBlock, agentID)
		delete(h.agentInfo, agentID)
	}
	delete(h.blockToAgent, blockID)

	log.Printf("[reactive] unregistered block %s", blockID)
}

// UpdateLastSeen updates the last seen time for an agent.
func (h *Handler) UpdateLastSeen(agentID string) {
	h.mu.Lock()
	defer h.mu.Unlock()

	if info, ok := h.agentInfo[agentID]; ok {
		info.LastSeen = time.Now()
	}
}

// GetAgent returns the registration info for an agent.
func (h *Handler) GetAgent(agentID string) *AgentRegistration {
	h.mu.RLock()
	defer h.mu.RUnlock()

	if info, ok := h.agentInfo[agentID]; ok {
		// Return a copy
		copy := *info
		return &copy
	}
	return nil
}

// GetAgentByBlock returns the agent ID for a block.
func (h *Handler) GetAgentByBlock(blockID string) string {
	h.mu.RLock()
	defer h.mu.RUnlock()
	return h.blockToAgent[blockID]
}

// ListAgents returns all registered agents.
func (h *Handler) ListAgents() []AgentRegistration {
	h.mu.RLock()
	defer h.mu.RUnlock()

	result := make([]AgentRegistration, 0, len(h.agentInfo))
	for _, info := range h.agentInfo {
		result = append(result, *info)
	}
	return result
}

// InjectMessage sends a message to a target agent's terminal.
func (h *Handler) InjectMessage(req InjectionRequest) InjectionResponse {
	// Generate request ID if not provided
	if req.RequestID == "" {
		req.RequestID = uuid.New().String()
	}

	// Rate limit check (prevents DoS while keeping synchronous injection)
	if !h.checkRateLimit() {
		return h.errorResponse(req, "rate limited: too many injection requests")
	}

	// Validate target agent ID
	if !ValidateAgentID(req.TargetAgentID) {
		return h.errorResponse(req, "invalid target agent ID")
	}

	// Sanitize the message
	sanitizedMsg := SanitizeMessage(req.Message)
	if sanitizedMsg == "" {
		return h.errorResponse(req, "message is empty after sanitization")
	}

	// Format message with optional source attribution
	finalMsg := FormatInjectedMessage(sanitizedMsg, req.SourceAgent, h.includeSourceInMessage)

	// Look up target block
	h.mu.RLock()
	blockID, exists := h.agentToBlock[req.TargetAgentID]
	h.mu.RUnlock()

	if !exists {
		return h.errorResponse(req, fmt.Sprintf("agent %s not found or not in a WaveMux pane", req.TargetAgentID))
	}

	// Send input to the block's PTY
	if h.inputSender == nil {
		return h.errorResponse(req, "input sender not configured")
	}

	// IMPORTANT: Message and Enter MUST be sent separately, not atomically.
	// Atomic writes (message + \r) do NOT reliably trigger input processing.
	// The PTY needs to see the message first, then Enter as a distinct event.
	//
	// ALSO IMPORTANT: This MUST be synchronous, not in a goroutine.
	// Async Enter sending was tried and failed - the Enter keys fire but
	// don't submit the message (they create blank lines instead).
	// The synchronous approach ensures message and Enter stay coordinated.

	// Step 1: Send the message content
	err := h.inputSender(blockID, []byte(finalMsg))
	if err != nil {
		h.logAudit(req, blockID, len(finalMsg), false, err.Error())
		return h.errorResponse(req, fmt.Sprintf("failed to send input: %v", err))
	}

	// Step 2: Small delay to let the terminal process the message
	time.Sleep(150 * time.Millisecond)

	// Step 3: Send Enter key (carriage return only, not CRLF)
	err = h.inputSender(blockID, []byte("\r"))
	if err != nil {
		// Log but don't fail - message was delivered, Enter just didn't work
		log.Printf("[reactive] Enter key send failed for block %s: %v", blockID, err)
	}

	// Log successful injection
	h.logAudit(req, blockID, len(finalMsg), true, "")

	return InjectionResponse{
		Success:   true,
		RequestID: req.RequestID,
		BlockID:   blockID,
		Timestamp: time.Now(),
	}
}

// errorResponse creates an error response and logs it.
func (h *Handler) errorResponse(req InjectionRequest, errMsg string) InjectionResponse {
	h.logAudit(req, "", len(req.Message), false, errMsg)
	return InjectionResponse{
		Success:   false,
		RequestID: req.RequestID,
		Error:     errMsg,
		Timestamp: time.Now(),
	}
}

// logAudit records an injection attempt to the audit log.
func (h *Handler) logAudit(req InjectionRequest, blockID string, msgLen int, success bool, errMsg string) {
	h.mu.Lock()
	defer h.mu.Unlock()

	// Compute message hash (for audit without storing content)
	hash := sha256.Sum256([]byte(req.Message))

	entry := AuditLogEntry{
		Timestamp:     time.Now(),
		SourceAgent:   req.SourceAgent,
		TargetAgent:   req.TargetAgentID,
		BlockID:       blockID,
		MessageHash:   hex.EncodeToString(hash[:8]), // First 8 bytes only
		MessageLength: msgLen,
		Success:       success,
		ErrorMessage:  errMsg,
		RequestID:     req.RequestID,
	}

	h.auditLog[h.auditLogIdx] = entry
	h.auditLogIdx = (h.auditLogIdx + 1) % h.auditLogSize
}

// GetAuditLog returns recent audit entries.
func (h *Handler) GetAuditLog(limit int) []AuditLogEntry {
	h.mu.RLock()
	defer h.mu.RUnlock()

	if limit <= 0 || limit > h.auditLogSize {
		limit = h.auditLogSize
	}

	result := make([]AuditLogEntry, 0, limit)

	// Read entries in reverse order (most recent first)
	idx := (h.auditLogIdx - 1 + h.auditLogSize) % h.auditLogSize
	for i := 0; i < limit; i++ {
		entry := h.auditLog[idx]
		if entry.Timestamp.IsZero() {
			break // No more entries
		}
		result = append(result, entry)
		idx = (idx - 1 + h.auditLogSize) % h.auditLogSize
	}

	return result
}

// SyncAgentsFromBlocks scans all blocks and registers any with WAVEMUX_AGENT_ID set.
// This is useful on startup to recover agent registrations.
func (h *Handler) SyncAgentsFromBlocks(ctx context.Context) error {
	blocks, err := wstore.DBGetAllObjsByType[*waveobj.Block](ctx, "block")
	if err != nil {
		return fmt.Errorf("failed to get blocks: %w", err)
	}

	for _, block := range blocks {
		if block.Meta == nil {
			continue
		}

		// Check if block has agent env var set
		cmdEnv, ok := block.Meta["cmd:env"].(map[string]interface{})
		if !ok {
			continue
		}

		agentID, ok := cmdEnv["WAVEMUX_AGENT_ID"].(string)
		if !ok || agentID == "" {
			continue
		}

		// Get tab ID from parent
		tabID := ""
		if block.ParentORef != "" {
			// ParentORef format is "tab:<tabid>"
			if len(block.ParentORef) > 4 && block.ParentORef[:4] == "tab:" {
				tabID = block.ParentORef[4:]
			}
		}

		h.RegisterAgent(agentID, block.OID, tabID)
	}

	return nil
}

// Global handler instance
var globalHandler *Handler
var globalHandlerOnce sync.Once

// GetGlobalHandler returns the global reactive handler instance.
func GetGlobalHandler() *Handler {
	globalHandlerOnce.Do(func() {
		// Handler will be initialized with proper inputSender later
		globalHandler = NewHandler(nil)
	})
	return globalHandler
}

// InitGlobalHandler initializes the global handler with an input sender.
func InitGlobalHandler(inputSender InputSender) {
	handler := GetGlobalHandler()
	handler.mu.Lock()
	handler.inputSender = inputSender
	handler.mu.Unlock()
}

File: agentmuxsrv-rs/src/backend/reactive/handler.rs
// Copyright 2025, Command Line Inc.
// SPDX-License-Identifier: Apache-2.0

#![allow(dead_code)]

use std::collections::HashMap;
use std::sync::{Mutex, OnceLock};
use std::time::{Duration, Instant};

use super::sanitize::{format_injected_message, sanitize_message, validate_agent_id};
use super::types::*;
use super::{now_unix_millis, sha256_hex, AUDIT_LOG_MAX, RATE_LIMIT_MAX};

// ---- Rate Limiter ----

pub(super) struct RateLimiter {
    tokens: u32,
    max_tokens: u32,
    last_refill: Instant,
}

impl RateLimiter {
    pub(super) fn new(max_tokens: u32) -> Self {
        Self {
            tokens: max_tokens,
            max_tokens,
            last_refill: Instant::now(),
        }
    }

    pub(super) fn check(&mut self) -> bool {
        let now = Instant::now();
        let elapsed = now.duration_since(self.last_refill);
        if elapsed >= Duration::from_secs(1) {
            self.tokens = self.max_tokens;
            self.last_refill = now;
        }
        if self.tokens > 0 {
            self.tokens -= 1;
            true
        } else {
            false
        }
    }
}

// ---- Handler ----

/// Core reactive messaging handler.
///
/// Manages agent registrations, rate limiting, message injection,
/// and audit logging.
pub struct Handler {
    agent_to_block: HashMap<String, String>,
    block_to_agent: HashMap<String, String>,
    agent_info: HashMap<String, AgentRegistration>,
    input_sender: Option<InputSender>,
    audit_log: Vec<AuditLogEntry>,
    rate_limiter: RateLimiter,
    include_source_in_message: bool,
}

impl Handler {
    /// Create a new handler without an input sender.
    /// Call `set_input_sender` before injecting messages.
    pub fn new() -> Self {
        Self {
            agent_to_block: HashMap::new(),
            block_to_agent: HashMap::new(),
            agent_info: HashMap::new(),
            input_sender: None,
            audit_log: Vec::with_capacity(AUDIT_LOG_MAX),
            rate_limiter: RateLimiter::new(RATE_LIMIT_MAX),
            include_source_in_message: false,
        }
    }

    /// Set the input sender function for message injection.
    pub fn set_input_sender(&mut self, sender: InputSender) {
        self.input_sender = Some(sender);
    }

    /// Set whether to include source agent prefix in injected messages.
    pub fn set_include_source(&mut self, include: bool) {
        self.include_source_in_message = include;
    }

    /// Register an agent with a block.
    pub fn register_agent(
        &mut self,
        agent_id: &str,
        block_id: &str,
        tab_id: Option<&str>,
    ) -> Result<(), String> {
        if !validate_agent_id(agent_id) {
            return Err(format!("invalid agent ID: {}", agent_id));
        }

        // Remove existing registration for this agent
        if let Some(old_block) = self.agent_to_block.remove(agent_id) {
            self.block_to_agent.remove(&old_block);
        }

        // Remove existing registration for this block
        if let Some(old_agent) = self.block_to_agent.remove(block_id) {
            self.agent_to_block.remove(&old_agent);
            self.agent_info.remove(&old_agent);
        }

        let now = now_unix_millis();
        self.agent_to_block
            .insert(agent_id.to_string(), block_id.to_string());
        self.block_to_agent
            .insert(block_id.to_string(), agent_id.to_string());
        self.agent_info.insert(
            agent_id.to_string(),
            AgentRegistration {
                agent_id: agent_id.to_string(),
                block_id: block_id.to_string(),
                tab_id: tab_id.map(|s| s.to_string()),
                registered_at: now,
                last_seen: now,
            },
        );

        Ok(())
    }

    /// Unregister an agent.
    pub fn unregister_agent(&mut self, agent_id: &str) {
        if let Some(block_id) = self.agent_to_block.remove(agent_id) {
            self.block_to_agent.remove(&block_id);
        }
        self.agent_info.remove(agent_id);
    }

    /// Unregister by block ID.
    pub fn unregister_block(&mut self, block_id: &str) {
        if let Some(agent_id) = self.block_to_agent.remove(block_id) {
            self.agent_to_block.remove(&agent_id);
            self.agent_info.remove(&agent_id);
        }
    }

    /// Update the last_seen timestamp for an agent.
    pub fn update_last_seen(&mut self, agent_id: &str) {
        if let Some(info) = self.agent_info.get_mut(agent_id) {
            info.last_seen = now_unix_millis();
        }
    }

    /// Get agent registration by agent ID.
    pub fn get_agent(&self, agent_id: &str) -> Option<&AgentRegistration> {
        self.agent_info.get(agent_id)
    }

    /// Get agent registration by block ID.
    pub fn get_agent_by_block(&self, block_id: &str) -> Option<&AgentRegistration> {
        self.block_to_agent
            .get(block_id)
            .and_then(|agent_id| self.agent_info.get(agent_id))
    }

    /// List all registered agents.
    pub fn list_agents(&self) -> Vec<AgentRegistration> {
        self.agent_info.values().cloned().collect()
    }

    /// Inject a message into an agent's terminal.
    ///
    /// Sends `message\r` as a single payload (required for text display),
    /// then spawns 3 delayed `\r` sends at 200ms intervals as separate
    /// PTY writes to ensure submission. See `specs/jekt-inject-timing.md`.
    pub fn inject_message(&mut self, mut req: InjectionRequest) -> InjectionResponse {
        let now = now_unix_millis();

        // Generate request ID if missing
        if req.request_id.is_none() || req.request_id.as_deref() == Some("") {
            req.request_id = Some(uuid::Uuid::new_v4().to_string());
        }
        let request_id = req.request_id.clone().unwrap_or_default();

        // Rate limit check
        if !self.rate_limiter.check() {
            return InjectionResponse {
                success: false,
                request_id,
                block_id: None,
                error: Some("rate limit exceeded".to_string()),
                timestamp: now,
            };
        }

        // Validate agent ID
        if !validate_agent_id(&req.target_agent) {
            return InjectionResponse {
                success: false,
                request_id,
                block_id: None,
                error: Some(format!("invalid agent ID: {}", req.target_agent)),
                timestamp: now,
            };
        }

        // Sanitize message
        let sanitized = sanitize_message(&req.message);

        // Look up block ID
        let block_id = match self.agent_to_block.get(&req.target_agent) {
            Some(id) => id.clone(),
            None => {
                let err = format!("agent not found: {}", req.target_agent);
                self.log_audit(
                    req.source_agent.as_deref(),
                    &req.target_agent,
                    "",
                    &sanitized,
                    false,
                    Some(&err),
                    &request_id,
                );
                return InjectionResponse {
                    success: false,
                    request_id,
                    block_id: None,
                    error: Some(err),
                    timestamp: now,
                };
            }
        };

        // Format message with source prefix if configured
        let final_msg = format_injected_message(
            &sanitized,
            req.source_agent.as_deref(),
            self.include_source_in_message,
        );

        // Send message via input sender
        let sender = match &self.input_sender {
            Some(s) => s.clone(),
            None => {
                let err = "input sender not configured".to_string();
                self.log_audit(
                    req.source_agent.as_deref(),
                    &req.target_agent,
                    &block_id,
                    &sanitized,
                    false,
                    Some(&err),
                    &request_id,
                );
                return InjectionResponse {
                    success: false,
                    request_id,
                    block_id: Some(block_id),
                    error: Some(err),
                    timestamp: now,
                };
            }
        };

        // Jekt inject sequence (see specs/jekt-inject-timing.md):
        // 1. \r to clear any partial input on the line
        // 2. message\r as single payload (proven to display text — v0.31.122/125)
        // 3. Three delayed \r at 200ms intervals as separate PTY writes to submit
        let _ = sender(&block_id, b"\r");
        let payload = format!("{}\r", final_msg);
        tracing::info!(
            target_agent = %req.target_agent,
            block_id = %block_id,
            msg_len = payload.len(),
            "inject: sending payload to PTY"
        );
        if let Err(e) = sender(&block_id, payload.as_bytes()) {
            tracing::error!(
                target_agent = %req.target_agent,
                block_id = %block_id,
                error = %e,
                "inject: sender failed"
            );
            self.log_audit(
                req.source_agent.as_deref(),
                &req.target_agent,
                &block_id,
                &sanitized,
                false,
                Some(&e),
                &request_id,
            );
            return InjectionResponse {
                success: false,
                request_id,
                block_id: Some(block_id),
                error: Some(e),
                timestamp: now,
            };
        }

        // Spawn 3 delayed \r sends as separate PTY events to ensure submission.
        let sender_enter = sender.clone();
        let block_id_enter = block_id.clone();
        tokio::spawn(async move {
            tokio::time::sleep(std::time::Duration::from_millis(200)).await;
            let _ = sender_enter(&block_id_enter, b"\r");
            tokio::time::sleep(std::time::Duration::from_millis(200)).await;
            let _ = sender_enter(&block_id_enter, b"\r");
            tokio::time::sleep(std::time::Duration::from_millis(200)).await;
            let _ = sender_enter(&block_id_enter, b"\r");
        });

        // Success
        self.log_audit(
            req.source_agent.as_deref(),
            &req.target_agent,
            &block_id,
            &sanitized,
            true,
            None,
            &request_id,
        );

        InjectionResponse {
            success: true,
            request_id,
            block_id: Some(block_id),
            error: None,
            timestamp: now,
        }
    }

    /// Get audit log entries, most recent first.
    pub fn get_audit_log(&self, limit: usize) -> Vec<AuditLogEntry> {
        let start = if self.audit_log.len() > limit {
            self.audit_log.len() - limit
        } else {
            0
        };
        let mut entries: Vec<_> = self.audit_log[start..].to_vec();
        entries.reverse();
        entries
    }

    /// Add an entry to the audit ring buffer.
    #[allow(clippy::too_many_arguments)]
    pub(super) fn log_audit(
        &mut self,
        source_agent: Option<&str>,
        target_agent: &str,
        block_id: &str,
        message: &str,
        success: bool,
        error_message: Option<&str>,
        request_id: &str,
    ) {
        let entry = AuditLogEntry {
            timestamp: now_unix_millis(),
            source_agent: source_agent.map(|s| s.to_string()),
            target_agent: target_agent.to_string(),
            block_id: block_id.to_string(),
            message_hash: sha256_hex(message),
            message_length: message.len(),
            success,
            error_message: error_message.map(|s| s.to_string()),
            request_id: request_id.to_string(),
        };

        if self.audit_log.len() >= AUDIT_LOG_MAX {
            self.audit_log.remove(0);
        }
        self.audit_log.push(entry);
    }
}

impl Default for Handler {
    fn default() -> Self {
        Self::new()
    }
}

// ---- Thread-safe wrapper ----

/// Thread-safe wrapper around Handler.
pub struct ReactiveHandler {
    inner: Mutex<Handler>,
}

impl ReactiveHandler {
    pub fn new() -> Self {
        Self {
            inner: Mutex::new(Handler::new()),
        }
    }

    pub fn set_input_sender(&self, sender: InputSender) {
        self.inner.lock().unwrap().set_input_sender(sender);
    }

    pub fn set_include_source(&self, include: bool) {
        self.inner.lock().unwrap().set_include_source(include);
    }

    pub fn register_agent(
        &self,
        agent_id: &str,
        block_id: &str,
        tab_id: Option<&str>,
    ) -> Result<(), String> {
        self.inner
            .lock()
            .unwrap()
            .register_agent(agent_id, block_id, tab_id)
    }

    pub fn unregister_agent(&self, agent_id: &str) {
        self.inner.lock().unwrap().unregister_agent(agent_id);
    }

    pub fn unregister_block(&self, block_id: &str) {
        self.inner.lock().unwrap().unregister_block(block_id);
    }

    pub fn update_last_seen(&self, agent_id: &str) {
        self.inner.lock().unwrap().update_last_seen(agent_id);
    }

    pub fn get_agent(&self, agent_id: &str) -> Option<AgentRegistration> {
        self.inner.lock().unwrap().get_agent(agent_id).cloned()
    }

    pub fn get_agent_by_block(&self, block_id: &str) -> Option<AgentRegistration> {
        self.inner
            .lock()
            .unwrap()
            .get_agent_by_block(block_id)
            .cloned()
    }

    pub fn list_agents(&self) -> Vec<AgentRegistration> {
        self.inner.lock().unwrap().list_agents()
    }

    pub fn inject_message(&self, req: InjectionRequest) -> InjectionResponse {
        self.inner.lock().unwrap().inject_message(req)
    }

    pub fn get_audit_log(&self, limit: usize) -> Vec<AuditLogEntry> {
        self.inner.lock().unwrap().get_audit_log(limit)
    }
}

impl Default for ReactiveHandler {
    fn default() -> Self {
        Self::new()
    }
}

/// Global reactive handler singleton.
static GLOBAL_HANDLER: OnceLock<ReactiveHandler> = OnceLock::new();

/// Get or initialize the global reactive handler.
pub fn get_global_handler() -> &'static ReactiveHandler {
    GLOBAL_HANDLER.get_or_init(ReactiveHandler::new)
}
